src/combinatorics/simple_graph.lean

```diff
@@ -0,0 +1,176 @@
+/-
+Copyright (c) 2020 Aaron Anderson, Jalex Stark, Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Aaron Anderson, Jalex Stark, Kyle Miller.
+-/
+import data.fintype.basic
+import data.sym2
+
+open finset
+
+/-!
+# Simple graphs
+
+This module defines simple graphs on a vertex type `V` as an
+irreflexive symmetric relation.
+
+There is a basic API for locally finite graphs and for graphs with
+finitely many vertices.
+
+- A locally finite graph is one with instances `∀ v, fintype (G.adj v)`.
+
+- Given instances `decidable_rel G.adj` and `fintype V`, then the graph
+is locally finite, too.
+
+## Implementation notes
+
+TODO: This is the simplest notion of an unoriented graph.  This should
+eventually fit into a more complete combinatorics hierarchy which
+includes multigraphs and directed graphs.  We begin with simple graphs
+in order to start learning what the combinatorics hierarchy should
+look like.
+
+TODO: Part of this would include defining, for example, subgraphs of a
+simple graph.
+
+-/
+
+universe u
+
+/--
+A simple graph is an irreflexive symmetric relation `adj` on a vertex type `V`.
+The relation describes which pairs of vertices are adjacent.
+There is exactly one edge for every pair of adjacent edges;
+see `simple_graph.E` for the corresponding type of edges.
+
+Note: The type of the relation is given as `V → set V` rather than
+`V → V → Prop` so that, given vertices `v` and `w`, then `w ∈ G.adj v`
+works as another way to write `G.adj v w`.  Otherwise Lean cannot find
+a `has_mem` instance.
+-/
+structure simple_graph (V : Type u) :=
+(adj : V → set V)
+(sym : symmetric adj . obviously)
+(loopless : irreflexive adj . obviously)
+
+/--
+The complete graph on a type `V` is the simple graph with all pairs of distinct vertices adjacent.
+-/
+def complete_graph (V : Type u) : simple_graph V :=
+{ adj := λ v w, v ≠ w }
+
+instance (V : Type u) : inhabited (simple_graph V) :=
+⟨complete_graph V⟩
+
+instance complete_graph_adj_decidable (V : Type u) [decidable_eq V] :
+  decidable_rel (complete_graph V).adj :=
+by { dsimp [complete_graph], apply_instance }
+
+namespace simple_graph
+variables {V : Type u} (G : simple_graph V)
+
+/--
+The edges of G consist of the unordered pairs of vertices related by
+`G.adj`.  It is given as a subtype of the symmetric square.
+-/
+def E : Type u := {x : sym2 V // x ∈ sym2.from_rel G.sym}
+
+/--
+`G.incident_edges v` is the set of edges incident to `v`.  Due to the way
+sets are defined, `G.incident_edges v e` denotes that `v` is incident to `e`.
```

Comment:
Please don't encourage people to break the set API. :-)

Reply:
I think it's being respected now...